### PR TITLE
[SPARK] CREATE EXTERNAL TABLE LOCATION

### PR DESCRIPTION
**Pull request by amaliujia**
**Description of changes**

Support Spark `CREATE EXTERNAL TABLE LOCATION` command

### CODE REVIEW

**Comment by amaliujia** [comment]
@cloud-fan

**Review by AnthonyByansi — APPROVED** [review]

File: connectors/spark/src/test/java/io/unitycatalog/connectors/spark/SparkIntegrationTest.java [hunk 1]
```diff
@@ -143,6 +145,24 @@ public void testCredentialDelta() throws ApiException, IOException {
...
+    String tableName = String.join(".", CATALOG_NAME, SCHEMA_NAME, "student_parquet");
+    session
+        .sql(
+            "CREATE TABLE "
+                + tableName
+                + "(name STRING) USING PARQUET LOCATION '"
+                + tempDirectory.toString()
+                + "'")
+        .collect();
+    assertTrue(session.catalog().tableExists(tableName));
```
Comment (cloud-fan) [inline connectors/spark/src/test/java/io/unitycatalog/connectors/spark/SparkIntegrationTest.java:162]: let's also test reading this empty table

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala [hunk 2]
```diff
@@ -191,10 +191,66 @@ private class UCProxy extends TableCatalog with SupportsNamespaces {
       .asInstanceOf[Table]
   }
 
-  override def createTable(ident: Identifier, columns: Array[Column], partitions: Array[Transform], properties: util.Map[String, String]): Table = ???
+  override def createTable(ident: Identifier, columns: Array[Column], partitions: Array[Transform], properties: util.Map[String, String]): Table = {
```
Comment (cloud-fan) [inline connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala:194]: we can simply remove this method here, as the interface provides a default implementation.

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala [hunk 3]
```diff
@@ -191,10 +191,66 @@ private class UCProxy extends TableCatalog with SupportsNamespaces {
...
+
 
   override def createTable(ident: Identifier, schema: StructType, partitions: Array[Transform], properties: util.Map[String, String]): Table = {
-    throw new AssertionError("deprecated `createTable` should not be called")
+    checkUnsupportedNestedNamespace(ident.namespace())
+    val createTable = new CreateTable()
+    createTable.setName(ident.name())
+    createTable.setSchemaName(ident.namespace().head)
+    createTable.setCatalogName(this.name)
+    val storageLocation = properties.get("location")
```
Comment (cloud-fan) [inline connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala:205]: let's use `TableCatalog.PROP_LOCATION`

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala [hunk 4]
```diff
@@ -191,10 +191,66 @@ private class UCProxy extends TableCatalog with SupportsNamespaces {
...
+      val column = new ColumnInfo()
+      column.setName(field.name)
+      if (field.getComment().isDefined) {
+        column.setComment(field.getComment.get)
+      }
+      column.setNullable(field.nullable)
+      column.setTypeText(field.dataType.simpleString)
+      column.setTypeName(convertColumnType(field.dataType))
+      // TODO: UC requires to set this value. Ideally this should be optional.
+      column.setTypeJson("")
```
Comment (cloud-fan) [inline connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala:223]: we can set it to `field.dataType.json`

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala [hunk 5]
```diff
@@ -191,10 +191,66 @@ private class UCProxy extends TableCatalog with SupportsNamespaces {
...
+      column.setPosition(i)
+      column
+    }
+    createTable.setColumns(columns)
+    assert(properties.get("provider") != null)
```
Comment (cloud-fan) [inline connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala:228]: please move all the checks to the beginning to fail earlier.

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala [hunk 6]
```diff
@@ -191,10 +191,66 @@ private class UCProxy extends TableCatalog with SupportsNamespaces {
...
+    val format: String = properties.get("provider")
+    createTable.setDataSourceFormat(convertDatasourceFormat(format))
+    tablesApi.createTable(createTable)
+    null
+  }
+
+  private def convertDatasourceFormat(format: String): DataSourceFormat = {
+    format match {
```
Comment (cloud-fan) [inline connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala:236]:
```suggestion
    format.toUpperCase match {
```

**Comment by amaliujia** [comment]
We need to wait Spark publish a RC including https://github.com/apache/spark/pull/47592 so that we can test this PR.

This is because Spark only support built-in file formats in `spark_catalog` where there is a bug that a custom session catalog does not work functionally before the fix.

File: connectors/spark/src/test/java/io/unitycatalog/connectors/spark/SparkIntegrationTest.java [hunk 7]
```diff
@@ -306,6 +306,24 @@ public void testLoadNamespace() throws IOException, ApiException {
     session.stop();
   }
 
+  @Test
+  public void testCreateExternalParquetTable() throws ApiException, IOException {
+    createCommonResources();
+    SparkSession session = createSparkSessionWithCatalogs(CATALOG_NAME);
```
Comment (cloud-fan) [inline connectors/spark/src/test/java/io/unitycatalog/connectors/spark/SparkIntegrationTest.java:312]: can we test with `SPARK_CATALOG`? then we can test SELECT *

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala [hunk 8]
```diff
@@ -191,10 +193,72 @@ private class UCProxy extends TableCatalog with SupportsNamespaces {
...
+    val columns: Seq[ColumnInfo] = schema.fields.toSeq.zipWithIndex.map { case (field, i) =>
+      val column = new ColumnInfo()
+      column.setName(field.name)
+      if (field.getComment().isDefined) {
+        column.setComment(field.getComment.get)
+      }
+      column.setNullable(field.nullable)
+      column.setTypeText(field.dataType.simpleString)
+      column.setTypeName(convertDataTypeToTypeName(field.dataType))
+      // TODO: UC requires to set this value. Ideally this should be optional.
```
Comment (cloud-fan) [inline connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala:222]:
```suggestion
```

File: connectors/spark/src/test/java/io/unitycatalog/connectors/spark/SparkIntegrationTest.java (hunk 7 above)
Comment (amaliujia) [reply connectors/spark/src/test/java/io/unitycatalog/connectors/spark/SparkIntegrationTest.java:312]: To remind you that `SPARK_CATALOG` does not work，it seems that even CreateTable does the DB exits check:


```
org.apache.spark.sql.catalyst.analysis.NoSuchDatabaseException: [SCHEMA_NOT_FOUND] The schema `uc_testschema` cannot be found. Verify the spelling and correctness of the schema and catalog.
If you did not qualify the name with a catalog, verify the current_schema() output, or qualify the name with the correct catalog.
To tolerate the error on drop use DROP SCHEMA IF EXISTS. SQLSTATE: 42704

	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.requireDbExists(SessionCatalog.scala:252)
	at org.apache.spark.sql.catalyst.catalog.SessionCatalog.createTable(SessionCatalog.scala:395)
```

This is because `SPARK_CATALOG` will always go to the spark built-in session catalog I think

Comment (amaliujia) [reply connectors/spark/src/test/java/io/unitycatalog/connectors/spark/SparkIntegrationTest.java:312]: hmmm if I use `default` schema then I have
```
org.apache.spark.sql.AnalysisException: [TABLE_OR_VIEW_NOT_FOUND] The table or view `spark_catalog`.`default`.`test_parquet` cannot be found. Verify the spelling and correctness of the schema and catalog.
If you did not qualify the name with a schema, verify the current_schema() output, or qualify the name with the correct schema and catalog.
To tolerate the error on drop use DROP VIEW IF EXISTS or DROP TABLE IF EXISTS. SQLSTATE: 42P01; line 1 pos 14;
'Project [*]
+- 'UnresolvedRelation [spark_catalog, default, test_parquet], [], false


	at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.tableNotFound(package.scala:87)
	at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis0$2(CheckAnalysis.scala:236)
	at org.apache.spark.sql.catalyst.analysis.CheckAnalysis.$anonfun$checkAnalysis0$2$adapted(CheckAnalysis.scala:216)
```

Basically `SPAKR_CATALOG` is not overwritten at all?

Comment (amaliujia) [reply connectors/spark/src/test/java/io/unitycatalog/connectors/spark/SparkIntegrationTest.java:312]: I am thinking if we need a lot more test and fix to make UC could overwrite the spark session catalog?

**Review by cloud-fan — APPROVED** [review]
